java/arcs/core/storage/StorageProxy.kt

```diff
@@ -53,87 +52,110 @@ class StorageProxy<Data : CrdtData, Op : CrdtOperationAtTime, T>(
 
     private val log = TaggedLog { "StorageProxy" }
     private val handleCallbacks = atomic(HandleCallbacks<T>())
-    private val stateHolder = atomic(StateHolder<T>(ProxyState.INIT))
+    private val stateHolder = atomic(StateHolder<T>(ProxyState.NO_SYNC))
     private val store: StorageCommunicationEndpoint<Data, Op, T> =
         storeEndpointProvider.getStorageEndpoint(ProxyCallback(::onMessage))
 
     /* visible for testing */
     fun getStateForTesting(): ProxyState = stateHolder.value.state
 
     /**
-     * Add a [Handle] `onReady` action, associated with a [Handle] name.
+     * If the [StorageProxy] is associated with any readable handles, it will need to operate
+     * in synchronized mode. This is done via a two-step process:
+     *   1) When constructed, all readable handles call this method to move the proxy from its
+     *      initial state of [NO_SYNC] to [READY_TO_SYNC].
+     *   2) [EntityHandleManager] then triggers the actual sync request after the arc has been
+     *      set up and all particles have received their onStart events.
+     */
+    fun prepareForSync() {
+        checkNotClosed()
+        stateHolder.update {
+            if (it.state == ProxyState.NO_SYNC) {
+                it.setState(ProxyState.READY_TO_SYNC)
+            } else {
+                it
+            }
+        }
+    }
+
+    /**
+     * If the [StorageProxy] has previously been set up for synchronized mode, send a sync request
+     * to the backing store and move to [AWAITING_SYNC].
+     */
+    fun maybeInitiateSync() {
+        checkNotClosed()
+        var needsSync = false
+        stateHolder.update {
+            // TODO(b/157188866): remove reliance on ready signal for write-only handles in tests
+            // If there are no readable handles observing this proxy, it will be in the NO_SYNC
+            // state and will never deliver any onReady notifications, which breaks tests that
+            // call awaitReady on write-only handles.
+            if (it.state == ProxyState.READY_TO_SYNC || it.state == ProxyState.NO_SYNC) {
+                needsSync = true
+                it.setState(ProxyState.AWAITING_SYNC)
+            } else {
+                it
```

Comment:
You'll want to re-set needsSync to false here, as `atomic.update`'s block is run until consensus is reached

Reply:
Can you explain this a bit more? I don't understand what the consensus is, and the code seems to work as it is. (I'm fine with adding the re-set, I'm just curious as to why it's actually needed.)

Reply:
Attempts by multiple threads to update an atomic "at the same time" using `update` will be re-run, comparing the before-state *before* running the update lambda against the before-state *after* running the update lambda -- and re-calling the lambda to calculate the new value based on the "current" value until that current value has been constant while the update lambda ran.

Here's the [source](https://github.com/Kotlin/kotlinx.atomicfu/blob/master/atomicfu/src/commonMain/kotlin/kotlinx/atomicfu/AtomicFU.common.kt#L100-L106)

Reply:
Thanks for the explanation, will fix.